src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -234,7 +256,33 @@ Checksum IChecksummedObject.Checksum
                 {
                     if (_lazyChecksum == null)
                     {
-                        _lazyChecksum = Checksum.Create(nameof(DocumentAttributes), this);
+                        using (var stream = SerializableBytes.CreateWritableStream())
+                        using (var writer = new ObjectWriter(stream))
+                        {
+                            writer.WriteString(nameof(DocumentAttributes));
+
+                            if (FilePath == null || ProjectFilePath == null)
+                            {
+                                // this checksum is not persistable because
+                                // this info doesn't have non volatile info
+                                Id.WriteTo(writer);
```

Comment:
checksum can be persistable and not persistable. but regardless of that, within same session, checksum must be consistent, deterministic and valid. having ID there make it consistent, deterministic. it just can't be persisted and re-used in different session.